File: core/src/main/scala/com/graphbrain/eco/nodes/WVRule.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.{Contexts, NodeType}

class WVRule(params: Array[ProgNode], lastTokenPos: Int= -1)
  extends RuleNode(params, lastTokenPos) {

  override val label = "wv"

  override def ntype = NodeType.Boolean

  override def vertexValue(ctxts: Contexts) = {
    // eval pattern
    params(0).booleanValue(ctxts)
    // eval conditions
    params(1).booleanValue(ctxts)
    // eval return value
    params(2).vertexValue(ctxts)

    for (c <- ctxts.ctxts)
      c.setRetVertex(this, c.getRetVertex(params(2)))
  }
}

File: core/src/main/java/com/graphbrain/eco/nodes/WVRule.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Context;
import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

public class WVRule extends RuleNode {

    public WVRule(ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
    }

    public WVRule(ProgNode[] params) {
        this(params, -1);
    }

    @Override
    public String label() {return "wv";}

    @Override
    public NodeType ntype(Context ctxt) {return NodeType.Boolean;}

    @Override
    public void eval(Contexts ctxts) {
        // eval pattern
        params[0].eval(ctxts);
        // eval conditions
        params[1].eval(ctxts);
        // eval return value
        params[2].eval(ctxts);

        for (Context c : ctxts.getCtxts())
            c.setRetVertex(this, c.getRetVertex(params[2]));
    }
}